In [1]:
import pandas as pd
import gurobipy as gp
from gurobipy import GRB
import scipy.sparse as sp
import numpy as np

In [15]:
brews_df = pd.read_pickle('data/brews_with_costs.pkl')
ingredients_df = pd.read_pickle('data/ingredients.pkl')
brews_df

first_ingredient       second_ingredient     third_ingredient  \
0      Abecean Longfin      Ancestor Moth Wing  Cyrodilic Spadetail   
1      Abecean Longfin       Ash Creep Cluster         Beehive Husk   
2      Abecean Longfin       Ash Creep Cluster         Chaurus Eggs   
3      Abecean Longfin       Ash Creep Cluster  Cyrodilic Spadetail   
4      Abecean Longfin       Ash Creep Cluster            Elves Ear   
...                ...                     ...                  ...   
25237    Tundra Cotton              Void Salts                   NA   
25238    Tundra Cotton          Wisp Wrappings                   NA   
25239     Vampire Dust               White Cap                   NA   
25240       Void Salts          Wisp Wrappings                   NA   
25241            Wheat  Yellow Mountain Flower                   NA   

                                      effects                            name  \
0       [Fortify Restoration, Damage Stamina]   Potion of Fortify Restoration   
1        [Fortify Sneak, Fortify Destruction]   Potion of Fortify Destruction   
2          [Weakness to Poison, Invisibility]          Potion of Invisibility   
3       [Fortify Restoration, Damage Stamina]   Potion of Fortify Restoration   
4            [Weakness to Frost, Resist Fire]           Potion of Resist Fire   
...                                       ...                             ...   
25237         [Resist Magic, Fortify Magicka]       Potion of Fortify Magicka   
25238                          [Resist Magic]          Potion of Resist Magic   
25239                       [Restore Magicka]       Potion of Restore Magicka   
25240                          [Resist Magic]          Potion of Resist Magic   
25241  [Fortify Health, Damage Stamina Regen]  Poison of Damage Stamina Regen   

       value                                       descriptions  
0        161  [Drain the target's Stamina by 18 points., Res...  
1        269  [Destruction spells are 30% stronger for 60 se...  
2        312  [Invisibility for 24 seconds., Target is 12% w...  
3        161  [Drain the target's Stamina by 18 points., Res...  
4        126  [Resist 18% of fire damage for 60 seconds., Ta...  
...      ...                                                ...  
25237    122  [Magicka is increased by 24 points for 60 seco...  
25238     51               [Resist 6% of magic for 60 seconds.]  
25239     25                    [Restore 30 points of Magicka.]  
25240     51               [Resist 6% of magic for 60 seconds.]  
25241    241  [Decrease the target's Stamina regeneration by...  

[25242 rows x 7 columns]

In [3]:
brews = list(brews_df.name)
ingredients = list(ingredients_df.name)

The model is the following:

#### Parameters

$B = \{1, \dots, 25242\}$: Set of brews

$I = \{1, \dots, 109\}$: Set of ingredients

$I_p \subset{I}$: Set of ingredients required to make one unit of brew $b$ (note that $|I_b| \leq 3 \,\,\forall b \in B$)

$z_i \in \mathbb{Z}^+$: Amount of ingredient $i$ available

$c_b \in \mathbb{R}^+$: gold value of brew $b$

#### Decision Variables

$y_b \in \mathbb{Z}^+$: Number of brews $b$ created

$x_{ib} \in \mathbb{Z}^+$: Amount of ingredient $i$ used to create brew $b$

#### Model

$$ \max \sum_{i \in b} c_b y_p$$
such that:
$$y_b \leq x_{ib} \qquad \forall i \in I_b,  \forall b \in B$$
$$\sum_{\{b \mid i \in I_b\}} x_{ib} \leq z_i \qquad \forall i \in I$$


An additional sophistication for the model is to allow the purchase of common ingredients, which would amount to adding a integer decision variable for each such ingredient, adding them to the RHSs of the second constraints, and subtracting the costs in the objective.

In [17]:
B = [i for i in range(len(brews_df))]
I = [i for i in range(len(ingredients_df))]
c = np.array(brews_df['value'])

In [10]:
A = [[] for b in B]
for b in B:
    brew_ingredients = brews_df.loc[b][['first_ingredient','second_ingredient','third_ingredient']]
    for ingredient in brew_ingredients:
        if ingredient != 'NA':
            i = ingredients.index(ingredient)
            A[b].append(i)
            
# randomly generate inventory
z = np.random.poisson(lam=2, size=len(I))

Using Gurobi:

In [18]:
l = gp.tuplelist()
for b in B:
    for i in A[b]:
        l.append((b,i))
        
m = gp.Model('alchemy')
x = m.addVars(l, vtype=GRB.CONTINUOUS, name='x')
y = m.addVars(B, vtype=GRB.INTEGER, name='y')
m.addConstrs((y[b] <= x[b,i] for (b,i) in l), 'c1')
m.addConstrs((x.sum('*', i) <= z[i] for i in I))

m.setObjective(gp.quicksum(c[b]*y[b] for b in B), GRB.MAXIMIZE)
m.setParam('OUTPUT_FLAG',False)
m.optimize()

In [19]:
optimal_brews = [b for b in B if y[b].x > 0.5]
optimal_df = brews_df.loc[optimal_brews]
optimal_df['count'] = [int(y[b].x) for b in B if y[b].x > 0.5]
optimal_df.sort_values(by=['value'], ascending=False)
optimal_df = optimal_df[['name', 'count', 'value', 'first_ingredient', 'second_ingredient', 'third_ingredient', 'descriptions']]
optimal_df['value_count'] = optimal_df['value'] * optimal_df['count']
optimal_df = optimal_df.sort_values(by='value_count', ascending=False)
optimal_df

name  count  value      first_ingredient  \
11281           Potion of Waterbreathing      1  24055         Chicken's Egg   
7183           Potion of Fortify Stamina      2   4023             Boar Tusk   
7126           Potion of Fortify Stamina      1   4116             Boar Tusk   
18939           Potion of Fortify Health      1   3029           Giant's Toe   
17724     Poison of Damage Stamina Regen      3    434         Frost Mirriam   
2618                 Poison of Paralysis      2    428       Ashen Grass Pod   
17032       Potion of Regenerate Magicka      2    374            Fire Salts   
10009       Potion of Fortify Pickpocket      2    347  Charred Skeever Hide   
22745       Potion of Regenerate Magicka      2    324         Jazbay Grapes   
13698       Potion of Fortify Two-handed      2    301       Dragon's Tongue   
19523     Poison of Damage Magicka Regen      1    584             Glow Dust   
19410                Poison of Paralysis      1    582          Gleamblossom   
8708      Poison of Damage Magicka Regen      1    559   Burnt Spriggan Wood   
22083                Poison of Paralysis      1    523           Human Flesh   
3138      Poison of Damage Magicka Regen      1    517            Bear Claws   
13276        Potion of Regenerate Health      1    508          Daedra Heart   
1796              Potion of Invisibility      1    506     Ash Creep Cluster   
25009  Poison of Lingering Damage Health      2    251            Scathecraw   
1041      Poison of Damage Magicka Regen      1    499    Ancestor Moth Wing   
1328              Potion of Invisibility      1    491     Ash Creep Cluster   
10585             Potion of Invisibility      1    490          Chaurus Eggs   
12837                     Poison of Slow      1    485   Cyrodilic Spadetail   
13733       Potion of Regenerate Stamina      1    475       Dragon's Tongue   
13602                     Poison of Slow      1    475             Deathbell   
9557                 Poison of Paralysis      1    467            Canis Root   
1242              Potion of Invisibility      1    467     Ash Creep Cluster   
1677      Poison of Damage Magicka Regen      1    459     Ash Creep Cluster   
5133              Potion of Invisibility      1    456           Blisterwort   
15111       Potion of Fortify Two-handed      1    455  Emperor Parasol Moss   
21512        Potion of Regenerate Health      1    436              Histcarp   
11280     Poison of Damage Magicka Regen      1    428         Chicken's Egg   
648       Poison of Damage Magicka Regen      1    426    Ancestor Moth Wing   
15218        Potion of Regenerate Health      1    411  Emperor Parasol Moss   
17702     Poison of Damage Stamina Regen      1    388         Frost Mirriam   
3110      Poison of Damage Magicka Regen      1    383            Bear Claws   
7831                 Poison of Paralysis      1    381           Briar Heart   
15433        Potion of Regenerate Health      1    360  Emperor Parasol Moss   
14259       Potion of Regenerate Magicka      1    347           Dwarven Oil   
18078       Potion of Regenerate Magicka      1    336           Frost Salts   
826       Poison of Damage Magicka Regen      1    330    Ancestor Moth Wing   
16433       Potion of Fortify Pickpocket      1    322            Falmer Ear   
11403     Poison of Damage Magicka Regen      1    308         Chicken's Egg   
1412              Potion of Invisibility      1    304     Ash Creep Cluster   
59         Potion of Fortify Restoration      1    287       Abecean Longfin   
14499     Poison of Damage Stamina Regen      1    255             Ectoplasm   
21788     Poison of Damage Stamina Regen      1    239             Honeycomb   
12855                     Poison of Fear      1    238   Cyrodilic Spadetail   
16241            Potion of Resist Poison      1    225            Falmer Ear   
9697             Potion of Resist Poison      1    225  Charred Skeever Hide   
18789             Potion of Resist S

In [20]:
print(f"Optimal Objective Value: {int(m.objVal)}")

Optimal Objective Value: 61143


In [21]:
optimal_df.loc[11281].descriptions

["Decrease the target's Magicka regeneration by 100% for 30 seconds.",
 "Drain the target's Stamina by 6 points per second for 10 seconds.",
 'Resist 6% of magic for 60 seconds.',
 'Can breathe underwater for 359 seconds.']